### ISSUE TITLE
Training loss is always NaN during training

### ISSUE THREAD

**Issue by oraby8** [MinishLab/tokenlearn#22]
i'm trying to use the same approach to train model in different dataset however  the training loss is always NaN during training.

![Image](https://github.com/user-attachments/assets/13903727-19a4-41c2-a2a0-ed428c9823d9)

**Comment by Pringled** [comment]
Hey @oraby8, this is likely due to the fact that we now quantize to `float16` by default in model2vec. In `train.py` ([here](https://github.com/MinishLab/tokenlearn/blob/main/tokenlearn/train.py#L39)), you can change quantize_to to "float32": `model = distill(model_name=model_name, quantize_to="float32"`. That should fix this. We will do a PR soon to fix this as well.

**Comment by oraby8** [comment]
Thanks @Pringled   for the quick response! I changed quantize_to to "float32" as suggested . it seemed to work initially, and the loss was decreasing. However, after a while, it started returning NaN again! Any ideas?

**Comment by Pringled** [comment]
I think it could be an issue with (very) small divisions in the the [sub_forward](https://github.com/MinishLab/tokenlearn/blob/main/tokenlearn/pretrain.py#L52) function. Could you try replacing that with the following:

```python
    def sub_forward(self, x: torch.Tensor) -> torch.Tensor:
        """Forward pass through the mean."""
        mask = (x != self.pad_id).float() 
        emb  = self.embeddings(x) * mask.unsqueeze(-1)
        weights = self.w[x] * mask
        sum_w = weights.sum(dim=1, keepdim=True).clamp(min=1.0)
        weighted_sum = (emb * weights.unsqueeze(-1)).sum(dim=1)
        return weighted_sum / sum_w   
```

If this works I can do a PR to fix both these issues. If not I'll dig a little deeper, thanks for flagging this!

**Comment by oraby8** [comment]
yes now it work. please do PR to fix it. Thanks
